plugin-locationcomponent/src/main/java/com/mapbox/maps/plugin/locationcomponent/LocationCompassEngine.kt

```diff
@@ -0,0 +1,321 @@
+        worldAxisForDeviceAxisY = SensorManager.AXIS_X
+      }
+      else -> {
+        worldAxisForDeviceAxisX = SensorManager.AXIS_X
+        worldAxisForDeviceAxisY = SensorManager.AXIS_Y
+      }
+    }
+    return Pair(worldAxisForDeviceAxisX, worldAxisForDeviceAxisY)
+  }
+
+  private fun notifyCompassChangeListeners(heading: Float) {
+    for (compassListener in compassListeners) {
+      compassListener.onCompassChanged(heading)
+    }
+    lastHeading = heading
+  }
+
+  private fun registerSensorListeners() {
+    if (isCompassSensorAvailable()) {
+      // Does nothing if the sensors already registered.
+      sensorManager.registerListener(
+        this,
+        compassSensor,
+        SENSOR_DELAY_MICROS
+      )
+    } else {
+      sensorManager.registerListener(
+        this,
+        gravitySensor,
+        SENSOR_DELAY_MICROS
+      )
+      sensorManager.registerListener(
+        this,
+        magneticFieldSensor,
+        SENSOR_DELAY_MICROS
+      )
+    }
+  }
+
+  private fun unregisterSensorListeners() {
+    if (isCompassSensorAvailable()) {
+      sensorManager.unregisterListener(this, compassSensor)
+    } else {
+      sensorManager.unregisterListener(this, gravitySensor)
+      sensorManager.unregisterListener(this, magneticFieldSensor)
+    }
+  }
+
+  private fun isCompassSensorAvailable() = compassSensor != null
+
+  /**
+   * Helper function, that filters newValues, considering previous values
+   *
+   * @param newValues      array of float, that contains new data
+   * @param smoothedValues array of float, that contains previous state
+   * @return float filtered array of float
+   */
+  private fun lowPassFilter(newValues: FloatArray, smoothedValues: FloatArray?): FloatArray {
+    if (smoothedValues == null) {
+      return newValues
+    }
+    for (i in newValues.indices) {
+      smoothedValues[i] =
+        smoothedValues[i] + ALPHA * (newValues[i] - smoothedValues[i])
+    }
+    return smoothedValues
+  }
+
+  /**
+   * Callback to receive compass update event
+   */
+  interface CompassListener {
+    /**
+     * Callback's invoked when a new compass update occurs.
+     *
+     * @param userHeading the new compass heading
+     */
+    fun onCompassChanged(userHeading: Float)
+  }
+
+  private companion object {
+    private const val TAG = "LocationCompassProvider"
+
+    // The rate sensor events will be delivered at. As the Android documentation states, this is only
+    // a hint to the system and the events might actually be received faster or slower then this
+    // specified rate. Since the minimum Android API levels about 9, we are able to set this value
+    // ourselves rather than using one of the provided constants which deliver updates too quickly for
+    // our use case. The default is set to 100ms
+    private const val SENSOR_DELAY_MICROS = 100 * 1000
+
+    // Filtering coefficient 0 < ALPHA < 1
+    private const val ALPHA = 0.45f
```

Comment:
Seems comment get lost - do we want do allow customize instead of hardcoding 0.45 ?